{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, collections"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing\n",
    "\n",
    "Depending on which transcriptome you used, you will need to create a file translating transcripts to genes. This notebook assumes the file is in `tr2g`, for ensembl transcriptomes these can be generated using biomart.\n",
    "\n",
    "The general format of `tr2g` is\n",
    "\n",
    "~~~\n",
    "ENST00000632684.1\tENSG00000282431.1\n",
    "ENST00000434970.2\tENSG00000237235.2\n",
    "ENST00000448914.1\tENSG00000228985.1\n",
    "ENST00000415118.1\tENSG00000223997.1\n",
    "ENST00000631435.1\tENSG00000282253.1\n",
    "ENST00000390583.1\tENSG00000211923.1\n",
    "ENST00000390577.1\tENSG00000211917.1\n",
    "ENST00000451044.1\tENSG00000227108.1\n",
    "ENST00000390578.1\tENSG00000211918.1\n",
    "ENST00000390572.1\tENSG00000211912.1\n",
    "...\n",
    "~~~\n",
    "\n",
    "The `matrix.ec` is generated by kallisto and connects the equivalence class ids to sets of transcripts. The format looks like\n",
    "\n",
    "~~~\n",
    "0\t0\n",
    "1\t1\n",
    "2\t2\n",
    "3\t3\n",
    "4\t4\n",
    "5\t5\n",
    "6\t6\n",
    "7\t7\n",
    "8\t8\n",
    "9\t9\n",
    "...\n",
    "884398\t26558,53383,53384,69915,69931,85319,109252,125730\n",
    "884399\t7750,35941,114698,119265\n",
    "884400\t9585,70083,92571,138545,138546\n",
    "884401\t90512,90513,134202,159456\n",
    "884402\t1460,2590,3476,8815,13440,13697,14581,18220,21799,21979,22752,23643,24268,31594,32066,32706,33922,33923,36234,40850,40851,49249,49250,50652,50989,51000,52533,54412,55545,56073,63432,63450,63457,63851,68524,70245,70679,70680,71800,77031,77202,77220,77867,92243,92749,92750,93617,96802,97437,98758,101367,109524,109838,113841,115937,116847,118451,118452,118984,118985,119680,121905,123916,125353,126987,127922,130003,130004,131999,132000,132001,132007,132008,132009,140512,140513,140514,144489,146571,150247,151565,151568,151571,151595,156609,156610,157764,157767,157773,157774,158578,166118,166119,169683,171534,171827,180225,184098\n",
    "884403\t83351,110474,110475,110476,110477,110478,110479,110480,110481,110482,174343\n",
    "884404\t22621,28045,37997,138681\n",
    "884405\t44088,44090,91875,91877,91878,91880,91884,91885,91887,91889,91894\n",
    "884406\t1232,1233,1234,1235,2387,2607,2608,7065,7066,7557,8638,8639,10905,11325,16266,21046,21047,21067,25410,25744,28980,29580,31453,31458,34886,35602,40319,45613,50352,55194,59781,62351,67708,71290,73284,76084,77412,77438,86173,86667,86669,86670,92243,110230,111434,116044,116847,123754,123756,123759,123892,123893,125169,129842,146226,146228,146571,149048,149949,151776,152059,153555,158139,160814,161642,162562,162591,165275,165276\n",
    "884407\t255,256,22621,33321,34387,37997,50333,53546,91651,96693,96706,121514,132702,138014,138681,140270,142596,169857\n",
    "~~~"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load transcript\n",
    "\n",
    "tr2g = {}\n",
    "trlist = []\n",
    "with open('data/tr2g') as f:\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        tr2g[l[0]] = l[1]\n",
    "        trlist.append(l[0])\n",
    "\n",
    "genes = list(set(tr2g[t] for t in tr2g))\n",
    "\n",
    "# load equivalence classes\n",
    "ecs = {}\n",
    "with open('data/matrix.ec') as f:\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        ec = int(l[0])\n",
    "        trs = [int(x) for x in l[1].split(',')]\n",
    "        ecs[ec] = trs\n",
    "        \n",
    "def ec2g(ec):\n",
    "    if ec in ecs:\n",
    "        return list(set(tr2g[trlist[t]] for t in ecs[ec]))        \n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processing the BUS file\n",
    "\n",
    "For this notebook we will work with the text file that `BUStools` produces, rather than the raw `BUS` file.\n",
    "\n",
    "We discard any barcodes that don't have more 10 UMIs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load test dataset\n",
    "\n",
    "cell_gene = collections.defaultdict(lambda: collections.defaultdict(float))\n",
    "pbar=None\n",
    "pumi=None\n",
    "with open('data/output.sorted.txt') as f:\n",
    "    gs = set()\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        barcode,umi,ec,count = line.split()\n",
    "        ec = int(ec)\n",
    "        \n",
    "        if barcode == pbar:\n",
    "            # same barcode\n",
    "            if umi == pumi:\n",
    "                # same UMI, let's update with intersection of genelist\n",
    "                gl = ec2g(ec)\n",
    "                gs.intersection_update(gl)\n",
    "            else:\n",
    "                # new UMI, process the previous gene set\n",
    "                for g in gs:\n",
    "                    cell_gene[barcode][g] += 1.0/len(gs)\n",
    "                # record new umi, reset gene set\n",
    "                pumi = umi\n",
    "                gs = set(ec2g(ec))\n",
    "        else:\n",
    "            # work with previous gene list\n",
    "            for g in gs:\n",
    "                cell_gene[pbar][g] += 1.0/len(gs)\n",
    "            \n",
    "            if sum(cell_gene[pbar][g] for g in cell_gene[pbar]) < 10:\n",
    "                del cell_gene[pbar]\n",
    "            \n",
    "            pbar = barcode\n",
    "            pumi = umi\n",
    "            \n",
    "            gs = set(ec2g(ec))\n",
    "    #remember the last gene\n",
    "    for g in gs:\n",
    "        cell_gene[pbar][g] += 1.0/len(gs)\n",
    "        \n",
    "    if sum(cell_gene[pbar][g] for g in cell_gene[pbar]) < 10:\n",
    "        del cell_gene[pbar]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_hist = collections.defaultdict(int)\n",
    "for barcode in cell_gene:\n",
    "    cg = cell_gene[barcode]\n",
    "    s = len([cg[g] for g in cg])\n",
    "    barcode_hist[barcode] += s\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "whitelist = set(x.strip() for x in open('10xwhitelist.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3609\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD8CAYAAABn919SAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAADt5JREFUeJzt3W+spOVZx/HvJX/bUt0FDmQDrGcxG+2+0EJOCIohDTS0gHExockao5tKsom2Sq1Gtjax9R0YbatJA1kLZjWkQLcYiFQroRDjC7fuluWfK+5CkW5Z2W3apa1pbLGXL+Y+eFzO7MycM3POea75fpLJzDzzDOe+5jn89pr7+XMiM5Ekdd+PrPYAJEnjYaBLUhEGuiQVYaBLUhEGuiQVYaBLUhEGuiQVYaBLUhEGuiQVcfpK/rDzzz8/Z2dnV/JHSlLn7d+//xuZOTNovRUN9NnZWfbt27eSP1KSOi8i/mOY9ZxykaQiDHRJKsJAl6QiDHRJKsJAl6QiDHRJKsJAl6QiDHRJKsJAl6QiVvRMUUmnNrvzkTcev3T7jas4EnWRHbokFWGgS1IRBrokFeEcujSlnK+vxw5dkoow0CWpCANdkoow0CWpCHeKSh3gDkwNww5dkoqwQ5eKs7ufHnboklSEHboku/gi7NAlqQg7dGmNWtg1S8OwQ5ekIgx0SSrCQJekIgx0SSrCQJekIgx0SSrCwxalKeKhkLUN1aFHxO9ExHMR8WxEfDYizo6ITRGxNyIORcT9EXHmpAcrSepvYIceERcBvw1syczvRcQDwDbgBuCTmXlfRNwF3ALcOdHRSkUs51R7T9NXP8POoZ8OvCUiTgfeChwFrgH2tNd3AzeNf3iSpGENDPTM/DrwJ8DL9IL8NWA/cCIzX2+rHQEumtQgJUmDDTPlsh7YCmwCTgCfA65fZNXs8/4dwA6AjRs3Lnmgkt7M6RctNMyUy7uBr2bm8cz8AfAg8HPAujYFA3Ax8Mpib87MXZk5l5lzMzMzYxm0JOnNhjls8WXgyoh4K/A94FpgH/A4cDNwH7AdeGhSg5Q0mIckapg59L30dn5+BXimvWcXcBvw4Yg4DJwH3D3BcUqSBhjqxKLM/BjwsZMWvwhcMfYRSZKWxFP/JakIT/2XVplz3xoXO3RJKsJAl6QinHKRJqiLJ/50cczqsUOXpCIMdEkqwkCXpCIMdEkqwkCXpCIMdEkqwkCXpCIMdEkqwkCXpCIMdEkqwlP/JQ3FSwKsfXboklSEgS5JRRjoklSEc+jSCuniXyZazpidc195duiSVISBLklFOOUiaVmcWlk77NAlqQgDXZKKMNAlqQjn0KUx6uKhiUsxLXV2jR26JBVhoEtSEQa6JBVhoEtSEQa6JBVhoEtSEQa6JBVhoEtSEQa6JBUxVKBHxLqI2BMR/xYRByPiZyPi3Ih4NCIOtfv1kx6sJKm/YTv0PwP+PjN/CvgZ4CCwE3gsMzcDj7XnkjTQ7M5H3rhpfAYGekT8KHA1cDdAZn4/M08AW4HdbbXdwE2TGqQkabBhOvRLgePAX0bEkxHxmYh4G3BhZh4FaPcXTHCckqQBhgn004HLgTsz8zLgvxhheiUidkTEvojYd/z48SUOU5I0yDCBfgQ4kpl72/M99AL+1YjYANDujy325szclZlzmTk3MzMzjjFLkhYx8HromfmfEfG1iPjJzHweuBb413bbDtze7h+a6EgldZY7P1fGsH/g4reAeyPiTOBF4P30uvsHIuIW4GXgfZMZoiRpGEMFemYeAOYWeena8Q5HkrRUnikqSUX4N0Uljc1S5soXvuel228cuFz92aFLUhEGuiQVYaBLUhEGuiQVYaBLUhEGuiQV4WGL0jJ5WrvWCjt0SSrCQJekIgx0SSrCQJekIgx0SSrCQJekIgx0SSrCQJekIgx0SSrCQJekIgx0SSrCa7lI+OfO1gqvi7M8duiSVIQdujQku3itdXboklSEHbq0BM71riy/HQ3HDl2SirBDl07BTlxdYocuSUUY6JJUhFMumlr9plOcZlFX2aFLUhEGuiQVYaBLUhHOoUvqFE8y6s8OXZKKsEPXVPEIlrrs3Efo0CPitIh4MiL+tj3fFBF7I+JQRNwfEWdObpiSpEFGmXK5FTi44PkdwCczczPwLeCWcQ5MkjSaoQI9Ii4GbgQ+054HcA2wp62yG7hpEgOUpFHN7nzkjds0GbZD/xTw+8AP2/PzgBOZ+Xp7fgS4aMxjkySNYOBO0Yj4BeBYZu6PiHfNL15k1ezz/h3ADoCNGzcucZjS0k1bl6bpNUyHfhXwixHxEnAfvamWTwHrImL+H4SLgVcWe3Nm7srMucycm5mZGcOQJUmLGRjomfmRzLw4M2eBbcCXMvNXgMeBm9tq24GHJjZKSdJAyzmx6DbgwxFxmN6c+t3jGZIkaSlGOrEoM58AnmiPXwSuGP+QJElL4an/klSEgS5JRXgtF0mlTdM1XuzQJakIO3RJneVJY/+fHbokFWGgS1IRBrokFWGgS1IRBrokFWGgS1IRBrokFWGgS1IRBrokFWGgS1IRnvovaWpUv1CXHbokFWGHrk7r13F50SZNIzt0SSrCQJekIpxykTT1quwstUOXpCIMdEkqwkCXpCKcQ1fn9Dsk0UMVNe3s0CWpCANdkoow0CWpCANdkopwp6jWlConeEirwQ5dkoqwQ5c0lSoe5mqHLklFGOiSVISBLklFOIcuSQsM81ew1uoRWAM79Ii4JCIej4iDEfFcRNzalp8bEY9GxKF2v37yw5Uk9TPMlMvrwO9m5juAK4EPRMQWYCfwWGZuBh5rzyVJq2TglEtmHgWOtsffiYiDwEXAVuBdbbXdwBPAbRMZpcrpwtdXqZ+1+vs70k7RiJgFLgP2Ahe2sJ8P/QvGPThJ0vCGDvSIOAf4PPChzPz2CO/bERH7ImLf8ePHlzJGSdIQhgr0iDiDXpjfm5kPtsWvRsSG9voG4Nhi783MXZk5l5lzMzMz4xizJGkRA+fQIyKAu4GDmfmJBS89DGwHbm/3D01khB20VufXVoOfhbqsa5cHGOY49KuAXwWeiYgDbdkf0AvyByLiFuBl4H2TGaIkaRjDHOXyT0D0efna8Q5HkrRUnvovSUUY6JJUhNdyGZE7+cZvmB1PXds5Ja0GO3RJKsIOvSMm8c1gOf/NkzvmSXxbsSuXRmOHLklFdL5DX+tz2sOMb1yd8mrWv1bGIU0zO3RJKqLzHfqohvlrJAtV6DZH7Z7ttqVuskOXpCIMdEkqojNTLpPeuTjMz51G4/xMp/2zlCbNDl2SiuhMh74cK9EZTuJnrGRH6+n30nitxsEFduiSVMRUdOiTMmrHOup+gK6qUIPURXboklREJzv0cXWAXe0kuzpuSZNlhy5JRRjoklREJ6dcKnDaRNK42aFLUhFlO3Q7YEkrYS1dndQOXZKKKNuhTyNP35dW12r//2WHLklFGOiSVISBLklFGOiSVISBLklFGOiSVISBLklFGOiSVISBLklFGOiSVISBLklFLCvQI+K9EfF8RByOiJ3jGpQkaXRLDvSIOA34NHA9sAX45YjYMq6BSZJGs5wO/QrgcGa+mJnfB+4Dto5nWJKkUS0n0C8Cvrbg+ZG2TJK0CpZzPfRYZFm+aaWIHcCO9vS7EfE8cD7wjWX87K6b5vqnuXaY7vqntva4A1he/T8+zErLCfQjwCULnl8MvHLySpm5C9i1cFlE7MvMuWX87E6b5vqnuXaY7vqnuXZYmfqXM+XyL8DmiNgUEWcC24CHxzMsSdKoltyhZ+brEfFB4IvAacA9mfnc2EYmSRrJsv6maGZ+AfjCEt66a/AqpU1z/dNcO0x3/dNcO6xA/ZH5pv2YkqQO8tR/SSpiYoEeES9FxDMRcSAi9rVl50bEoxFxqN2vb8sjIv68XULg6Yi4fFLjmoSIuCcijkXEswuWjVxrRGxv6x+KiO2rUctS9Kn/4xHx9bb9D0TEDQte+0ir//mIeM+C5Z27lEREXBIRj0fEwYh4LiJubcvLb/9T1D4t2/7siPhyRDzV6v+jtnxTROxt2/H+dtAIEXFWe364vT674L+16OcyssycyA14CTj/pGV/DOxsj3cCd7THNwB/R+/Y9iuBvZMa14RqvRq4HHh2qbUC5wIvtvv17fH61a5tGfV/HPi9RdbdAjwFnAVsAl6gt1P9tPb4UuDMts6W1a5tiNo3AJe3x28H/r3VWH77n6L2adn2AZzTHp8B7G3b9AFgW1t+F/Ab7fFvAne1x9uA+0/1uSxlTCs95bIV2N0e7wZuWrD8r7Lnn4F1EbFhhce2ZJn5j8A3T1o8aq3vAR7NzG9m5reAR4H3Tn70y9en/n62Avdl5n9n5leBw/QuI9HJS0lk5tHM/Ep7/B3gIL0zpstv/1PU3k+1bZ+Z+d329Ix2S+AaYE9bfvK2n/+d2ANcGxFB/89lZJMM9AT+ISL2R+9sUYALM/Mo9H4ZgAva8oqXERi11oqfwQfbtMI981MOFK6/fYW+jF6nNlXb/6TaYUq2fUScFhEHgGP0/hF+ATiRma+3VRbW8kad7fXXgPMYY/2TDPSrMvNyeldj/EBEXH2KdYe6jEAR/Wqt9hncCfwE8E7gKPCnbXnJ+iPiHODzwIcy89unWnWRZZ2uf5Hap2bbZ+b/ZOY76Z0pfwXwjsVWa/cTr39igZ6Zr7T7Y8Df0Cv21fmplHZ/rK0+1GUEOmbUWkt9Bpn5avtl/yHwF/zfV8hy9UfEGfQC7d7MfLAtnortv1jt07Tt52XmCeAJenPo6yJi/hyfhbW8UWd7/cfoTVWOrf6JBHpEvC0i3j7/GLgOeJbepQHm995vBx5qjx8Gfq0dAXAl8Nr819UOG7XWLwLXRcT69hX1urask07aB/JL9LY/9Orf1vb4bwI2A1+mo5eSaHOgdwMHM/MTC14qv/371T5F234mIta1x28B3k1vP8LjwM1ttZO3/fzvxM3Al7K3V7Tf5zK6Ce39vZTeXtungOeAj7bl5wGPAYfa/bkL9hZ/mt780zPA3CTGNakb8Fl6Xy1/QO9f21uWUivw6/R2iBwG3r/adS2z/r9u9T3dfmE3LFj/o63+54HrFyy/gd6REi/M/86s9Rvw8/S+Hj8NHGi3G6Zh+5+i9mnZ9j8NPNnqfBb4w7b8UnqBfBj4HHBWW352e364vX7poM9l1JtnikpSEZ4pKklFGOiSVISBLklFGOiSVISBLklFGOiSVISBLklFGOiSVMT/ArRP4AiMZrj3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "bcv = [x for b,x in barcode_hist.items() if x > 600 and x < 3000   and b in whitelist]\n",
    "_ = plt.hist(bcv,bins=100)\n",
    "print(len(bcv))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9375892828799086"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_barcode = [x for x in barcode_hist if  barcode_hist[x] <= 100]\n",
    "len(bad_barcode)/len(cell_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 0\n",
    "bad_s = 0\n",
    "bad_barcode_set = set(bad_barcode)\n",
    "for barcode in cell_gene:\n",
    "    cg = cell_gene[barcode]\n",
    "    cgs =  sum(cg[g] for g in cg)\n",
    "    s += cgs\n",
    "    if barcode in bad_barcode_set:\n",
    "        bad_s += cgs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.18355898685976038"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_s/s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to Seurat compatible matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = 'matrix.mtx'\n",
    "\n",
    "gene_to_id = dict((g,i+1) for i,g in enumerate(genes))\n",
    "barcodes_to_use = [b for b,x in barcode_hist.items() if x > 600 and x < 3000 and b in whitelist]\n",
    "\n",
    "num_entries = 0\n",
    "for barcode in barcodes_to_use:\n",
    "    num_entries += len([x for x in cell_gene[barcode].values() if round(x)>0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(outfile, 'w') as of:\n",
    "    of.write('%%MatrixMarket matrix coordinate real general\\n%\\n')\n",
    "    #number of genes\n",
    "    of.write(\"%d %d %d\\n\"%(len(genes), len(barcodes_to_use), num_entries))\n",
    "    bcid = 0\n",
    "    for barcode in barcodes_to_use:\n",
    "        bcid += 1\n",
    "        cg = cell_gene[barcode]\n",
    "        gl = [(gene_to_id[g],round(cg[g])) for g in cg if round(cg[g]) > 0]\n",
    "        gl.sort()\n",
    "        for x in gl:\n",
    "            of.write(\"%d %d %d\\n\"%(x[0],bcid,x[1]))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = {}\n",
    "with open('mart_export.txt') as f:\n",
    "    f.readline()\n",
    "    for line in f:\n",
    "        g,t,gn = line.split()\n",
    "        gene_names[g] = gn\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_genes = dict((i,g) for (g,i) in gene_to_id.items())\n",
    "gl = []\n",
    "for i in range(1,len(genes)+1):\n",
    "    g = id_to_genes[i]\n",
    "    gid = g[:g.find('.')]\n",
    "    if gid in gene_names:\n",
    "        gn = gene_names[gid]\n",
    "    else:\n",
    "        gn = ''\n",
    "    gl.append((g,gn))\n",
    "\n",
    "with open('genes.tsv','w') as of:\n",
    "    for g,gn in gl:\n",
    "        of.write(\"%s\\t%s\\n\"%(g,gn))\n",
    "        \n",
    "with open('barcodes.tsv','w') as of:\n",
    "    of.write('\\n'.join(x + '-1' for x in barcodes_to_use))\n",
    "    of.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gl[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
